File: examples/sidecar-channel-node/adapter.js
#!/usr/bin/env node
/**
 * Example sidecar channel adapter for LibreFang (Node.js)
 *
 * Usage in config.toml:
 *   [[sidecar_channels]]
 *   name = "node-echo"
 *   command = "node"
 *   args = ["examples/sidecar-channel-node/adapter.js"]
 */

const readline = require("readline");

function sendEvent(method, params) {
  const event = { method };
  if (params) event.params = params;
  process.stdout.write(JSON.stringify(event) + "\n");
}

function handleCommand(cmd) {
  if (!cmd || typeof cmd !== "object" || Array.isArray(cmd) || typeof cmd.method !== "string") {
    sendEvent("error", { message: "Invalid command: expected an object with a string method" });
    return;
  }

  switch (cmd.method) {
    case "send": {
      if (!cmd.params || typeof cmd.params !== "object" || Array.isArray(cmd.params)) {
        sendEvent("error", { message: "Invalid send params: expected an object" });
        return;
      }
      sendEvent("message", {
        user_id: "echo-user",
        user_name: "Echo Bot (Node)",
        text: `Echo: ${cmd.params?.text || ""}`,
        channel_id: cmd.params?.channel_id || "default",
      });
      break;
    }
    case "shutdown":
      process.exit(0);
      break;
    default:
      sendEvent("error", { message: `Unsupported command: ${cmd.method}` });
  }
}

// Signal readiness
sendEvent("ready");

// Read commands from stdin
const rl = readline.createInterface({ input: process.stdin });
rl.on("line", (line) => {
  try {
    handleCommand(JSON.parse(line));
  } catch (e) {
    sendEvent("error", { message: `Invalid JSON: ${e.message}` });
  }
});

File: examples/sidecar-channel-go/adapter.go
// Example sidecar channel adapter for LibreFang (Go)
//
// Build:
//
//	go build -o adapter adapter.go
//
// Usage in config.toml:
//
//	[[sidecar_channels]]
//	name = "go-echo"
//	command = "./examples/sidecar-channel-go/adapter"
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
)

type Event struct {
	Method string      `json:"method"`
	Params interface{} `json:"params,omitempty"`
}

type Command struct {
	Method string          `json:"method"`
	Params json.RawMessage `json:"params,omitempty"`
}

type SendParams struct {
	ChannelID string `json:"channel_id"`
	Text      string `json:"text"`
}

type MessageParams struct {
	UserID    string `json:"user_id"`
	UserName  string `json:"user_name"`
	Text      string `json:"text"`
	ChannelID string `json:"channel_id"`
}

func sendEvent(method string, params interface{}) error {
	evt := Event{Method: method, Params: params}
	return json.NewEncoder(os.Stdout).Encode(evt)
}

func main() {
	// Signal readiness
	if err := sendEvent("ready", nil); err != nil {
		fmt.Fprintf(os.Stderr, "encode ready event: %v\n", err)
		return
	}

	// Read commands from stdin
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			continue
		}

		var cmd Command
		if err := json.Unmarshal([]byte(line), &cmd); err != nil {
			_ = sendEvent("error", map[string]string{"message": fmt.Sprintf("Invalid JSON: %v", err)})
			continue
		}

		switch cmd.Method {
		case "send":
			var params SendParams
			if err := json.Unmarshal(cmd.Params, &params); err != nil {
				_ = sendEvent("error", map[string]string{"message": fmt.Sprintf("Invalid send params: %v", err)})
				continue
			}
			if err := sendEvent("message", MessageParams{
				UserID:    "echo-user",
				UserName:  "Echo Bot (Go)",
				Text:      fmt.Sprintf("Echo: %s", params.Text),
				ChannelID: params.ChannelID,
			}); err != nil {
				fmt.Fprintf(os.Stderr, "encode message event: %v\n", err)
				return
			}
		case "shutdown":
			return
		default:
			_ = sendEvent("error", map[string]string{"message": fmt.Sprintf("Unsupported command: %s", cmd.Method)})
		}
	}
	if err := scanner.Err(); err != nil {
		fmt.Fprintf(os.Stderr, "read command: %v\n", err)
	}
}
